contracts/src/registry/NameWrapperRegistry.sol

```diff
@@ -0,0 +1,247 @@
+    }
+
+    function uri(uint256 /*tokenId*/ ) public pure override returns (string memory) {
+        return "";
+    }
+
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
+    {
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(oldFlags);
+        if (expires < block.timestamp) {
+            return address(0);
+        }
+        return super.ownerOf(tokenId);
+    }
+
+    function register(string calldata label, address owner, IRegistry registry, uint96 flags, uint64 expires)
+        public
+        onlyRole(ROOT_CONTEXT, REGISTRAR_ROLE)
+        returns (uint256 tokenId)
+    {
+        tokenId = _canonicalTokenId(uint256(keccak256(bytes(label))), flags);
+        flags = (flags & FLAGS_MASK) | (uint96(expires) << 32);
+
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiry = _extractExpiry(oldFlags);
+        if (oldExpiry >= block.timestamp) {
+            revert NameAlreadyRegistered(label);
+        }
+
+        // if there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        datastore.setSubregistry(tokenId, address(registry), flags);
+
+        // registrar has some permissions by default
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _grantRole(tokenIdContext, RENEW_ROLE, msg.sender);
+        _grantRole(tokenIdContext, EMANCIPATOR_ROLE, msg.sender);
+
+        emit NewSubname(label);
+        return tokenId;
+    }
+
+    function renew(uint256 tokenId, uint64 expires) public onlyRole(_tokenIdContext(tokenId), RENEW_ROLE) {
+        address sender = _msgSender();
+
+        (address subregistry, uint96 flags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiration = _extractExpiry(flags);
+        if (oldExpiration < block.timestamp) {
+            revert NameExpired(tokenId);
+        }
+        if (expires < oldExpiration) {
+            revert CannotReduceExpiration(oldExpiration, expires);
+        }
+        datastore.setSubregistry(tokenId, subregistry, (flags & FLAGS_MASK) | (uint96(expires) << 32));
+
+        emit NameRenewed(tokenId, expires, sender);
+    }
+
+    /** 
+     * @dev Emancipate the owner of the token. 
+     * 
+     * See v1 fuse: PARENT_CANNOT_CONTROL
+     */
+    function emancipate(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATOR_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _grantRole(tokenIdContext, EMANCIPATED_OWNER_ROLE, ownerOf(tokenId));
+        renounceRole(tokenIdContext, EMANCIPATOR_ROLE, msg.sender); // remove the emancipator role since we can only do this once.
+    }
```

Comment:
I freely admit I haven't thought all the implications of this through, but do we need to have this as an explicit thing? Can't we just revoke our own control over the name while giving the same control to the owner of the subname?

Reply:
I did consider role-transference but this would violate the rule that a role can only be assigned someone who is allowed to assign it (e.g admin can assign registrar role). If a user can transfer their role to someone else then they are effectively able to bypass that.